### PR TITLE
fix(Exact): support array union

### LINKED ISSUE #419: `Exact` intersection type support
I was playing with @zorji 's `Exact` type and found that it errors for intersection types, but specifically when you construct the intersection type like `Array<{ a: string }> & Array<{ b: string }>`

I can only reproduce the error in the following scenario:

```typescript
type Type = Array<{ test: string }> & Array<{ z: number }>
// Can't construct Exact type with errors saying it's trying to also match on array method 'pop' and etc etc
const fn = <T extends Exact<Type, T>>(args: T) => args;

// correct input
fn([{
  code: 'test',
  name: '1'
}])

// missing field
fn([{
  code: 'test'
}])

// incorrect type
fn([{
  code: 'test',
  name: 1
}])

// excess field
fn([{
  code: 'test',
  name: '1',
  excessField: '',
}])
```

As per https://www.reddit.com/r/typescript/comments/m5wvrx/how_to_extract_proper_intersection_type_from_an/ and https://github.com/microsoft/TypeScript/issues/43267 

It seems like a workaround for those who have control over type declarations would be writing the intersection type like following:

```typescript
type Type = ({ code: string } & { name: string })[]
const fn = <T extends Exact<Type, T>>(args: T) => args;

// correct input
fn([{
  code: 'test',
  name: '1'
}])

// missing field
fn([{
  code: 'test'
}])

// incorrect type
fn([{
  code: 'test',
  name: 1
}])

// excess field
fn([{
  code: 'test',
  name: '1',
  excessField: '',
}])
```

The only problem is when the user doesn't have control over the underlying module's type declaration but they are providing the weird intersection type that goes `Array<{ ... }> & Array<{...}>`, then they will get the error when trying to use `Exact`

### PR DESCRIPTION
**Pull request by zorji**
Fixes #419

### CODE REVIEW

**Comment by zorji** [comment]
@JasonShin could you help review this PR? including grammar issues lol

File: test-d/exact.ts [hunk 1]
```diff
@@ -143,3 +143,69 @@ import type {Exact} from '../index';
...
+			y: '',
+			z: 1,
+		}];
+		// @ts-expect-error
+		fn(input);
+	}
+}
+
+{ // Spec - union of array with nested fields
+	type Type = Array<{x: string}> & Array<{z: number; d: {e: string}}>;
```
Comment (JasonShin) [inline test-d/exact.ts:196]: Thanks @zorji !! I was testing this again and found that union of 2 ReadOnlyArray breaks the type. But I think creating a union of 2 read-only arrays is already broken in TS... 

Screen Shot 2022-07-11 at 9 09 11 am

Unless you think this is also important to handle in this PR 👍

File: test-d/exact.ts [hunk 2]
```diff
@@ -143,3 +143,69 @@ import type {Exact} from '../index';
...
+			x: '',
+			z: 1,
+			d: {
+				e: 'test',
+				f: true, // Excess field
+			},
+		}];
+		// @ts-expect-error
+		fn(input);
+	}
```
Comment (JasonShin) [inline test-d/exact.ts:210]: Possibly we can include more test coverages to check excess, exact and missing fields for the union array test

File: source/exact.d.ts [hunk 3]
```diff
@@ -41,11 +61,8 @@ onlyAcceptNameImproved(invalidInput); // Compilation error
...
-export type Exact<ParameterType, InputType extends ParameterType> = ParameterType extends Primitive
-	? ParameterType
-	/*
-	Create a type from `ParameterType` and `InputType` and change keys exclusive to `InputType` to `never`.
-	- Generate a list of keys that exists in `InputType` but not in `ParameterType`.
-	- Mark these excess keys as `never`.
-	*/
-	: {[Key in keyof ParameterType]: Exact<ParameterType[Key], InputType[Key]>} & Record<Exclude<keyof InputType, KeysOfUnion<ParameterType>>, never>;
+export type Exact<ParameterType, InputType> =
+	// Converting union of array to array of union. i.e. A[] & B[] => (A & B)[]
```
Comment (JasonShin) [inline source/exact.d.ts:65]: 👍 
Screen Shot 2022-07-11 at 9 13 55 am

**Comment by JasonShin** [comment]
Otherwise looks good, I will leave grammar and spelling checks to others 😊

File: test-d/exact.ts [hunk 4]
```diff
@@ -143,3 +143,69 @@ import type {Exact} from '../index';
 		fn(input);
 	}
 }
+
```
Comment (JasonShin) [inline test-d/exact.ts:146]: A little bit out of the scope of this PR but can we also include exhaustive tests against ReadonlyArray as well? I just want to ensure ReadonlyArray will work consistently in the future along with the regular mutable arrays.

Screen Shot 2022-07-11 at 9 22 56 am

We can just copy & paste Array spec for the ReadonlyArray test

**Comment by zorji** [comment]
@JasonShin looks like ReadOnly is an easy fix but I am not sure whether this is a breaking change. I'll put them in 2 separate PRs. https://github.com/zorji/type-fest/pull/1/files

**Comment by zorji** [comment]
@JasonShin actually I think I found a way

**Review by JasonShin — APPROVED** [review]
Looks good from my end, thanks for the updates!

**Comment by JasonShin** [comment]
@sindresorhus Please review this thanks 👍

File: source/exact.d.ts [hunk 5]
```diff
@@ -1,6 +1,26 @@
-import type {Primitive} from './primitive';
 import type {KeysOfUnion} from './internal';
 
+/**
+Extract the element of an array that also works for array union.
+Return `never` if T is not an array.
+It creates a type-safe way to access the element type of `unknown` type.
+*/
+type ArrayElement<T> = T extends readonly unknown[]? T[0] : never;
```
Comment (sindresorhus) [inline source/exact.d.ts:8]: This should be moved to internal.d.ts

**Comment by zorji** [comment]
> Could this use the trick mentioned in https://github.com/sindresorhus/type-fest/issues/118#issuecomment-657259814 ?

@sindresorhus I have tested and found the suggested implementation doesn't work for union of array

**Pull request merged**